cats/cats-redis/src/test/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentSchedulerTest.java

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.cats.redis.cluster;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyList;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.netflix.spinnaker.cats.agent.CachingAgent;
+import com.netflix.spinnaker.cats.agent.DefaultCacheResult;
+import com.netflix.spinnaker.cats.agent.ExecutionInstrumentation;
+import com.netflix.spinnaker.cats.cluster.DefaultAgentIntervalProvider;
+import com.netflix.spinnaker.cats.test.TestAgent;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.junit.Before;
+import org.junit.Test;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+
+public class ClusteredSortAgentSchedulerTest {
+
+  private ClusteredSortAgentScheduler clusteredSortAgentScheduler;
+
+  private Jedis jedis = mock(Jedis.class);
+  private JedisPool jedisPool = mock(JedisPool.class);
+  private Integer parallelism = 2;
+  private CachingAgent.CacheExecution agentExecution = mock(CachingAgent.CacheExecution.class);
+  private ExecutionInstrumentation executionInstrumentation = mock(ExecutionInstrumentation.class);
+
+  private Optional<Semaphore> runningAgents;
+
+  @Before
+  public void setUp() throws IllegalAccessException {
+    when(jedisPool.getResource()).thenReturn(jedis);
+    when(jedis.scriptLoad(anyString())).thenReturn("testScriptSha");
+    when(jedis.time()).thenReturn(List.of("1678784468", "374338"));
+    DefaultAgentIntervalProvider intervalProvider = new DefaultAgentIntervalProvider(6000000);
+    clusteredSortAgentScheduler =
+        new ClusteredSortAgentScheduler(jedisPool, () -> false, intervalProvider, parallelism);
+
+    runningAgents =
+        (Optional<Semaphore>)
+            FieldUtils.getDeclaredField(ClusteredSortAgentScheduler.class, "runningAgents", true)
+                .get(clusteredSortAgentScheduler);
+  }
+
+  @Test
+  public void testRunningAgentsSemaphore() {
+    when(jedis.zrangeByScore(eq(ClusteredSortAgentScheduler.WORKING_SET), anyString(), anyString()))
+        .thenReturn(new HashSet<>());
+    when(jedis.zrangeByScore(eq(ClusteredSortAgentScheduler.WAITING_SET), anyString(), anyString()))
+        .thenReturn(Set.of("testAgentType"));
+
+    clusteredSortAgentScheduler.saturatePool();
+
+    // When parallelism is -1, the runningAgents is Optional.empty().
```

Comment:
But isn't parallelism initialized to 2 above?

Reply:
At first I thought that if parallelism is -1, both test cases will also pass. But think about it, these two test cases are used to test Semaphore's permit leaking, if parallelism is -1, there is no Semaphore.